File: scripts/validate_r4_structures.py
#!/usr/bin/env python3
"""Validate MVP profiles and generated resources against supplied FHIR R4 StructureDefinitions.

This is an offline, deterministic structural check. It deliberately uses the
R4 StructureDefinitions as the source of truth rather than the package's
OpenAPI JSON schemas, which are not complete profile validators.
"""
from __future__ import annotations

import argparse
import json
import tarfile
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable

ROOT = Path(__file__).resolve().parents[1]
RES = ROOT / "fsh-generated" / "resources"
FHIR_CANONICAL_PREFIX = "http://hl7.org/fhir/StructureDefinition/"

PRIMITIVE_TYPES = {
    "base64Binary", "boolean", "canonical", "code", "date", "dateTime",
    "decimal", "id", "instant", "integer", "markdown", "oid",
    "positiveInt", "string", "time", "unsignedInt", "uri", "url", "uuid",
}
INTEGER_TYPES = {"integer", "positiveInt", "unsignedInt"}
STRING_TYPES = PRIMITIVE_TYPES - {"boolean", "decimal"} - INTEGER_TYPES


@dataclass
class PackageReader:
    path: Path

    def __post_init__(self) -> None:
        self._tar: tarfile.TarFile | None = None
        self._prefix = ""
        if self.path.is_file():
            self._tar = tarfile.open(self.path, "r:gz")
            names = set(self._tar.getnames())
            self._prefix = "package/" if "package/package.json" in names else ""
        elif not self.path.is_dir():
            raise FileNotFoundError(self.path)

    def read_json(self, filename: str) -> dict[str, Any]:
        if self._tar is not None:
            member = self._tar.extractfile(self._prefix + filename)
            if member is None:
                raise FileNotFoundError(filename)
            return json.load(member)
        candidates = [self.path / filename, self.path / "package" / filename]
        for candidate in candidates:
            if candidate.exists():
                return json.loads(candidate.read_text(encoding="utf-8"))
        raise FileNotFoundError(filename)

    def close(self) -> None:
        if self._tar is not None:
            self._tar.close()


def max_value(value: str | None) -> float:
    if value in (None, "*"):
        return float("inf")
    return float(value)


def type_suffix(code: str) -> str:
    return code[0].upper() + code[1:]


def value_matches_type(value: Any, code: str) -> bool:
    if code == "boolean":
        return isinstance(value, bool)
    if code in INTEGER_TYPES:
        return isinstance(value, int) and not isinstance(value, bool)
    if code == "decimal":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if code in STRING_TYPES:
        return isinstance(value, str)
    return isinstance(value, dict)


def load_generated() -> tuple[list[dict[str, Any]], dict[str, dict[str, Any]]]:
    resources: list[dict[str, Any]] = []
    by_url: dict[str, dict[str, Any]] = {}
    for path in sorted(RES.glob("*.json")):
        obj = json.loads(path.read_text(encoding="utf-8"))
        obj["__filename"] = path.name
        resources.append(obj)
        if obj.get("url"):
            by_url[obj["url"]] = obj
    return resources, by_url


def base_type_for_profile(url: str, generated_by_url: dict[str, dict[str, Any]]) -> str | None:
    if url.startswith(FHIR_CANONICAL_PREFIX):
        return url.removeprefix(FHIR_CANONICAL_PREFIX)
    profile = generated_by_url.get(url)
    if not profile:
        return None
    return profile.get("type")


def validate_profile_differential(
    profile: dict[str, Any],
    core: PackageReader,
    generated_by_url: dict[str, dict[str, Any]],
) -> list[str]:
    errors: list[str] = []
    base_url = profile.get("baseDefinition")
    if not base_url or not base_url.startswith(FHIR_CANONICAL_PREFIX):
        return [f"{profile['__filename']}: unsupported baseDefinition {base_url!r}"]
    base_type = base_url.removeprefix(FHIR_CANONICAL_PREFIX)
    try:
        base = core.read_json(f"StructureDefinition-{base_type}.json")
    except FileNotFoundError:
        return [f"{profile['__filename']}: missing core StructureDefinition for {base_type}"]
    base_elements = {e["path"]: e for e in base["snapshot"]["element"]}
    for element in profile.get("differential", {}).get("element", []):
        path = element["path"]
        if path not in base_elements:
            errors.append(f"{profile['__filename']}: differential path not in R4 base: {path}")
            continue
        parent = base_elements[path]
        if element.get("min", parent.get("min", 0)) < parent.get("min", 0):
            errors.append(f"{profile['__filename']}: {path} lowers min cardinality")
        if max_value(element.get("max", parent.get("max"))) > max_value(parent.get("max")):
            errors.append(f"{profile['__filename']}: {path} raises max cardinality")
        if "type" in element:
            base_types = {t["code"] for t in parent.get("type", [])}
            for narrowed in element["type"]:
                code = narrowed["code"]
                if code not in base_types:
                    errors.append(
                        f"{profile['__filename']}: {path} type {code} is not allowed by R4 base {sorted(base_types)}"
                    )
                if code == "Reference" and narrowed.get("targetProfile"):
                    allowed_targets = {
                        target
                        for t in parent.get("type", [])
                        if t.get("code") == "Reference"
                        for target in t.get("targetProfile", [])
                    }
                    allowed_types = {base_type_for_profile(t, generated_by_url) for t in allowed_targets}
                    for target in narrowed["targetProfile"]:
                        target_type = base_type_for_profile(target, generated_by_url)
                        if allowed_targets and target not in allowed_targets and target_type not in allowed_types:
                            errors.append(
                                f"{profile['__filename']}: {path} targetProfile {target} is not a subtype of an allowed R4 target"
                            )
    return errors


def child_elements(sd: dict[str, Any]) -> dict[str, dict[str, Any]]:
    root = sd["type"]
    result: dict[str, dict[str, Any]] = {}
    for element in sd["snapshot"]["element"]:
        path = element["path"]
        if not path.startswith(root + "."):
            continue
        tail = path[len(root) + 1 :]
        if "." not in tail:
            result[tail] = element
    return result


def validate_resource_top_level(resource: dict[str, Any], core: PackageReader) -> list[str]:
    errors: list[str] = []
    rt = resource.get("resourceType")
    filename = resource.get("__filename", f"{rt}/{resource.get('id', '?')}")
    if not rt:
        return [f"{filename}: resourceType missing"]
    try:
        sd = core.read_json(f"StructureDefinition-{rt}.json")
    except FileNotFoundError:
        return [f"{filename}: no R4 StructureDefinition found for {rt}"]
    elements = child_elements(sd)
    allowed_keys = {"resourceType", "__filename"}
    choice_keys: dict[str, str] = {}
    for name, element in elements.items():
        if name.endswith("[x]"):
            base_name = name[:-3]
            for t in element.get("type", []):
                key = base_name + type_suffix(t["code"])
                allowed_keys.add(key)
                choice_keys[key] = t["code"]
        else:
            allowed_keys.add(name)
            if any(t.get("code") in PRIMITIVE_TYPES for t in element.get("type", [])):
                allowed_keys.add("_" + name)
    for key in resource:
        if key not in allowed_keys:
            errors.append(f"{filename}: unknown top-level R4 element {key}")
    for name, element in elements.items():
        min_card = element.get("min", 0)
        max_card = element.get("max", "1")
        if name.endswith("[x]"):
            base_name = name[:-3]
            present = [k for k in resource if k.startswith(base_name) and k in choice_keys]
            if min_card and not present:
                errors.append(f"{filename}: required choice {name} is missing")
            if len(present) > 1:
                errors.append(f"{filename}: more than one choice supplied for {name}: {present}")
            for key in present:
                if not value_matches_type(resource[key], choice_keys[key]):
                    errors.append(f"{filename}: {key} has wrong JSON type for {choice_keys[key]}")
            continue
        if min_card and name not in resource:
            errors.append(f"{filename}: required R4 element {name} is missing")
        if name not in resource:
            continue
        value = resource[name]
        if max_card == "*" or (max_card.isdigit() and int(max_card) > 1):
            if not isinstance(value, list):
                errors.append(f"{filename}: repeating element {name} must be a JSON array")
                continue
            values = value
        else:
            if isinstance(value, list):
                errors.append(f"{filename}: singleton element {name} must not be a JSON array")
                continue
            values = [value]
        types = [t["code"] for t in element.get("type", [])]
        if types:
            for item in values:
                if not any(value_matches_type(item, code) for code in types):
                    errors.append(f"{filename}: {name} has wrong JSON type for R4 types {types}")
    return errors


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--r4-package",
        type=Path,
        required=True,
        help="Path to hl7.fhir.r4.core#4.0.1 package directory or .tgz",
    )
    args = parser.parse_args()
    core = PackageReader(args.r4_package)
    resources, generated_by_url = load_generated()
    errors: list[str] = []
    profile_count = 0
    instance_count = 0
    for resource in resources:
        errors.extend(validate_resource_top_level(resource, core))
        if resource.get("resourceType") == "StructureDefinition" and resource.get("derivation") == "constraint":
            profile_count += 1
            errors.extend(validate_profile_differential(resource, core, generated_by_url))
        else:
            instance_count += 1
        if resource.get("resourceType") == "Bundle":
            for entry in resource.get("entry", []):
                nested = dict(entry.get("resource", {}))
                if nested:
                    nested["__filename"] = f"{resource['__filename']}::{nested.get('resourceType')}/{nested.get('id', '?')}"
                    errors.extend(validate_resource_top_level(nested, core))
    core.close()

    report = ROOT / "validation" / "r4-structure-validation.md"
    report.parent.mkdir(exist_ok=True)
    lines = [
        "# FHIR R4 StructureDefinition validation",
        "",
        f"Validated {profile_count} constrained profiles and {instance_count} other generated resources against the supplied FHIR R4 4.0.1 StructureDefinitions.",
        "",
    ]
    if errors:
        lines += ["## Errors", ""] + [f"- {e}" for e in errors]
    else:
        lines += [
            "**PASS** — all differential paths, cardinality restrictions, type restrictions, and generated resource top-level structures are compatible with FHIR R4 4.0.1.",
            "",
            "This deterministic offline check is complementary to, not a replacement for, the HL7 FHIR Validator and IG Publisher QA.",
        ]
    report.write_text("\n".join(lines) + "\n", encoding="utf-8")
    print(report.read_text(encoding="utf-8"), end="")
    return 1 if errors else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate-r4-structures.ts
#!/usr/bin/env bun
/**
 * Validate MVP profiles and generated resources against supplied FHIR R4
 * StructureDefinitions. This deterministic offline check is complementary to
 * the HL7 Validator and IG Publisher.
 */

import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { basename, dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));
const RES = join(ROOT, "fsh-generated", "resources");
const FHIR_CANONICAL_PREFIX = "http://hl7.org/fhir/StructureDefinition/";
const PRIMITIVE_TYPES = new Set(["base64Binary", "boolean", "canonical", "code", "date", "dateTime", "decimal", "id", "instant", "integer", "markdown", "oid", "positiveInt", "string", "time", "unsignedInt", "uri", "url", "uuid"]);
const FHIRPATH_TYPE_MAP: Record<string, string> = {
  "http://hl7.org/fhirpath/System.Boolean": "boolean",
  "http://hl7.org/fhirpath/System.Decimal": "decimal",
  "http://hl7.org/fhirpath/System.Integer": "integer",
  "http://hl7.org/fhirpath/System.String": "string",
};
const INTEGER_TYPES = new Set(["integer", "positiveInt", "unsignedInt"]);
const STRING_TYPES = new Set([...PRIMITIVE_TYPES].filter((x) => x !== "boolean" && x !== "decimal" && !INTEGER_TYPES.has(x)));

const glob = (pattern: string, cwd: string) => Array.from(new Bun.Glob(pattern).scanSync({ cwd })).sort();
const exists = async (path: string) => Bun.file(path).exists();
const loadJson = async (path: string) => JSON.parse(await Bun.file(path).text());
const normalizeType = (code: string) => FHIRPATH_TYPE_MAP[code] || code;
const maxValue = (value?: string) => value == null || value === "*" ? Infinity : Number(value);
const typeSuffix = (code: string) => {
  const normalized = normalizeType(code);
  return normalized[0].toUpperCase() + normalized.slice(1);
};

class PackageReader {
  root: string;
  tempRoot?: string;

  constructor(path: string) {
    if (path.endsWith(".tgz") || path.endsWith(".tar.gz")) {
      this.tempRoot = mkdtempSync(join(tmpdir(), "ptmvp-r4-"));
      const proc = Bun.spawnSync(["tar", "-xzf", path, "-C", this.tempRoot]);
      if (!proc.success) throw new Error(`could not extract ${path}: ${new TextDecoder().decode(proc.stderr)}`);
      this.root = this.tempRoot;
    } else {
      this.root = path;
    }
  }

  async readJson(filename: string) {
    for (const candidate of [join(this.root, filename), join(this.root, "package", filename)]) {
      if (await exists(candidate)) return loadJson(candidate);
    }
    throw new Error(`missing ${filename}`);
  }

  close() {
    if (this.tempRoot) rmSync(this.tempRoot, { recursive: true, force: true });
  }
}

function valueMatchesType(value: any, code: string) {
  const normalized = normalizeType(code);
  if (normalized === "boolean") return typeof value === "boolean";
  if (INTEGER_TYPES.has(normalized)) return Number.isInteger(value);
  if (normalized === "decimal") return typeof value === "number";
  if (STRING_TYPES.has(normalized)) return typeof value === "string";
  return value && typeof value === "object" && !Array.isArray(value);
}

async function loadGenerated() {
  const resources: any[] = [];
  const byUrl = new Map<string, any>();
  for (const file of glob("*.json", RES)) {
    const obj = await loadJson(join(RES, file));
    obj.__filename = file;
    resources.push(obj);
    if (obj.url) byUrl.set(obj.url, obj);
  }
  return { resources, byUrl };
}

function baseTypeForProfile(url: string, generatedByUrl: Map<string, any>) {
  if (url.startsWith(FHIR_CANONICAL_PREFIX)) return url.slice(FHIR_CANONICAL_PREFIX.length);
  return generatedByUrl.get(url)?.type || null;
}

async function validateProfileDifferential(profile: any, core: PackageReader, generatedByUrl: Map<string, any>) {
  const errors: string[] = [];
  const baseUrl = profile.baseDefinition;
  let base: any;
  if (baseUrl?.startsWith(FHIR_CANONICAL_PREFIX)) {
    const baseType = baseUrl.slice(FHIR_CANONICAL_PREFIX.length);
    try {
      base = await core.readJson(`StructureDefinition-${baseType}.json`);
    } catch {
      return [`${profile.__filename}: missing core StructureDefinition for ${baseType}`];
    }
  } else {
    base = generatedByUrl.get(baseUrl);
    if (!base) return [`${profile.__filename}: unsupported baseDefinition ${JSON.stringify(baseUrl)}`];
    if (!base.snapshot?.element && base.type) {
      try {
        base = await core.readJson(`StructureDefinition-${base.type}.json`);
      } catch {
        return [`${profile.__filename}: local baseDefinition ${baseUrl} has no snapshot and no core StructureDefinition for ${base.type}`];
      }
    }
  }
  const baseElements = new Map(base.snapshot.element.map((e: any) => [e.path, e]));
  for (const element of profile.differential?.element || []) {
    const parent: any = baseElements.get(element.path);
    if (!parent) {
      errors.push(`${profile.__filename}: differential path not in R4 base: ${element.path}`);
      continue;
    }
    if ((element.min ?? parent.min ?? 0) < (parent.min ?? 0)) errors.push(`${profile.__filename}: ${element.path} lowers min cardinality`);
    if (maxValue(element.max ?? parent.max) > maxValue(parent.max)) errors.push(`${profile.__filename}: ${element.path} raises max cardinality`);
    if (element.type) {
      const baseTypes = new Set((parent.type || []).map((t: any) => t.code));
      for (const narrowed of element.type) {
        if (!baseTypes.has(narrowed.code)) errors.push(`${profile.__filename}: ${element.path} type ${narrowed.code} is not allowed by R4 base ${JSON.stringify([...baseTypes].sort())}`);
        if (narrowed.code === "Reference" && narrowed.targetProfile) {
          const allowedTargets = new Set<string>();
          for (const t of parent.type || []) if (t.code === "Reference") for (const target of t.targetProfile || []) allowedTargets.add(target);
          const allowedTypes = new Set([...allowedTargets].map((t) => baseTypeForProfile(t, generatedByUrl)));
          for (const target of narrowed.targetProfile) {
            const targetType = baseTypeForProfile(target, generatedByUrl);
            if (allowedTargets.size && !allowedTargets.has(target) && !allowedTypes.has(targetType)) {
              errors.push(`${profile.__filename}: ${element.path} targetProfile ${target} is not a subtype of an allowed R4 target`);
            }
          }
        }
      }
    }
  }
  return errors;
}

function childElements(sd: any) {
  const root = sd.type;
  const result = new Map<string, any>();
  for (const element of sd.snapshot.element) {
    const path = element.path;
    if (!path.startsWith(`${root}.`)) continue;
    const tail = path.slice(root.length + 1);
    if (!tail.includes(".")) result.set(tail, element);
  }
  return result;
}

async function validateResourceTopLevel(resource: any, core: PackageReader) {
  const errors: string[] = [];
  const rt = resource.resourceType;
  const filename = resource.__filename || `${rt}/${resource.id || "?"}`;
  if (!rt) return [`${filename}: resourceType missing`];
  let sd: any;
  try {
    sd = await core.readJson(`StructureDefinition-${rt}.json`);
  } catch {
    return [`${filename}: no R4 StructureDefinition found for ${rt}`];
  }
  const elements = childElements(sd);
  const allowedKeys = new Set(["resourceType", "__filename"]);
  const choiceKeys = new Map<string, string>();
  for (const [name, element] of elements) {
    if (name.endsWith("[x]")) {
      const baseName = name.slice(0, -3);
      for (const type of element.type || []) {
        const key = baseName + typeSuffix(type.code);
        allowedKeys.add(key);
        choiceKeys.set(key, type.code);
      }
    } else {
      allowedKeys.add(name);
      if ((element.type || []).some((t: any) => PRIMITIVE_TYPES.has(normalizeType(t.code)))) allowedKeys.add(`_${name}`);
    }
  }
  for (const key of Object.keys(resource)) {
    if (!allowedKeys.has(key)) errors.push(`${filename}: unknown top-level R4 element ${key}`);
  }
  for (const [name, element] of elements) {
    const min = element.min || 0;
    const max = element.max || "1";
    if (name.endsWith("[x]")) {
      const baseName = name.slice(0, -3);
      const present = Object.keys(resource).filter((k) => k.startsWith(baseName) && choiceKeys.has(k));
      if (min && !present.length) errors.push(`${filename}: required choice ${name} is missing`);
      if (present.length > 1) errors.push(`${filename}: more than one choice supplied for ${name}: ${JSON.stringify(present)}`);
      for (const key of present) if (!valueMatchesType(resource[key], choiceKeys.get(key)!)) errors.push(`${filename}: ${key} has wrong JSON type for ${choiceKeys.get(key)}`);
      continue;
    }
    if (min && !(name in resource)) errors.push(`${filename}: required R4 element ${name} is missing`);
    if (!(name in resource)) continue;
    const value = resource[name];
    const values = max === "*" || (Number.isInteger(Number(max)) && Number(max) > 1) ? value : [value];
    if ((max === "*" || (Number.isInteger(Number(max)) && Number(max) > 1)) && !Array.isArray(value)) {
      errors.push(`${filename}: repeating element ${name} must be a JSON array`);
      continue;
    }
    if (!(max === "*" || (Number.isInteger(Number(max)) && Number(max) > 1)) && Array.isArray(value)) {
      errors.push(`${filename}: singleton element ${name} must not be a JSON array`);
      continue;
    }
    const types = (element.type || []).map((t: any) => t.code);
    for (const item of values) {
      if (types.length && !types.some((code: string) => valueMatchesType(item, code))) errors.push(`${filename}: ${name} has wrong JSON type for R4 types ${JSON.stringify(types)}`);
    }
  }
  return errors;
}

function argValue(name: string) {
  const idx = Bun.argv.indexOf(name);
  return idx >= 0 ? Bun.argv[idx + 1] : null;
}

async function main() {
  const r4Package = argValue("--r4-package");
  if (!r4Package) {
    console.error(`Usage: bun ${basename(Bun.argv[1])} --r4-package /path/to/hl7.fhir.r4.core#4.0.1`);
    return 2;
  }
  const core = new PackageReader(r4Package);
  const { resources, byUrl } = await loadGenerated();
  const errors: string[] = [];
  let profileCount = 0;
  let instanceCount = 0;
  try {
    for (const resource of resources) {
      errors.push(...await validateResourceTopLevel(resource, core));
      if (resource.resourceType === "StructureDefinition" && resource.derivation === "constraint") {
        profileCount++;
        errors.push(...await validateProfileDifferential(resource, core, byUrl));
      } else {
        instanceCount++;
      }
      if (resource.resourceType === "Bundle") {
        for (const entry of resource.entry || []) {
          if (!entry.resource) continue;
          const nested = { ...entry.resource, __filename: `${resource.__filename}::${entry.resource.resourceType}/${entry.resource.id || "?"}` };
          errors.push(...await validateResourceTopLevel(nested, core));
        }
      }
    }
  } finally {
    core.close();
  }

  const lines = [
    "# FHIR R4 StructureDefinition validation",
    "",
    `Validated ${profileCount} constrained profiles and ${instanceCount} other generated resources against the supplied FHIR R4 4.0.1 StructureDefinitions.`,
    "",
    ...(errors.length
      ? ["## Errors", "", ...errors.map((e) => `- ${e}`)]
      : ["**PASS** - all differential paths, cardinality restrictions, type restrictions, and generated resource top-level structures are compatible with FHIR R4 4.0.1.", "", "This deterministic offline check is complementary to, not a replacement for, the HL7 FHIR Validator and IG Publisher QA."]),
  ];
  const report = join(ROOT, "validation", "r4-structure-validation.md");
  await Bun.write(report, lines.join("\n") + "\n");
  process.stdout.write(await Bun.file(report).text());
  return errors.length ? 1 : 0;
}

process.exit(await main());
